### ISSUE TITLE
Replace `MinDistance(TreeType*)` with `MinDistance(TreeType&)`

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#787]
Right now, the TreeType API requires the following functions to have this signature:

```
ElemType MinDistance(const TreeType* other);
ElemType MaxDistance(const TreeType* other);
math::RangeType<ElemType> RangeDistance(const TreeType* other);
```

But we can easily use references here.  We should refactor the trees (`mlpack/core/tree/*/`) to instead use the signature

```
ElemType MinDistance(const TreeType& other);
ElemType MaxDistance(const TreeType& other);
math::RangeType<ElemType> RangeDistance(const TreeType& other);
```

That will also force some changes in the dual-tree algorithm code, but it should be relatively simple refactoring in the end.  This would be a good starter bug for someone looking to get involved---but I want it to be done before the 2.1.0 release, which is soon, so I may end up doing this myself in the next two weeks if nobody beats me to a PR. :)

**Comment by chvsp** [comment]
Hi @rcurtin I would like to solve this bug. Seems pretty straightforward. I would be starting work after 22 September as I have exams till then. Just letting you know.

**Comment by rcurtin** [comment]
Fixed by #788.  Thanks @akanuraj200!
